File: peas_backend/peas_app/api/mappers/recipe_mapper.py
from typing import List

from peas_app.api.dtos.v1.requests.api_requests import RecipeRequest
from peas_app.api.dtos.v1.responses.api_responses import (
    RecipeProductResponse,
    RecipeResponse,
)
from peas_app.api.mappers.product_mapper import ProductMapper
from peas_app.api.mappers.unit_mapper import UnitMapper
from peas_app.database.models.models import (
    Product,
    ProductRecipeAssociation,
    Recipe,
)


class RecipeMapper:
    @staticmethod
    def update_recipe(
        recipe_to_update: Recipe,
        recipe_request: RecipeRequest,
        products: List[Product],
    ) -> Recipe:
        recipe_to_update.name = recipe_request.name
        recipe_to_update.description = recipe_request.description
        recipe_to_update.servings = recipe_request.servings
        recipe_to_update.products = products
        return recipe_to_update

    @staticmethod
    def to_recipe_with_id(
        recipe_request: RecipeRequest, products: List[Product], recipe_id: int
    ) -> Recipe:
        return Recipe(
            id=recipe_id,
            name=recipe_request.name,
            description=recipe_request.description,
            servings=recipe_request.servings,
            products=products,
        )

    @staticmethod
    def to_recipe(recipe_request: RecipeRequest) -> Recipe:
        return Recipe(
            name=recipe_request.name,
            description=recipe_request.description,
            servings=recipe_request.servings,
            product_associations=RecipeMapper.to_product_associations(
                recipe_request
            ),
        )

    @staticmethod
    def to_product_associations(
        recipe_request: RecipeRequest,
    ) -> List[ProductRecipeAssociation]:
        return [
            ProductRecipeAssociation(
                product_id=product.id,
                quantity=product.quantity,
                unit=UnitMapper.to_sql_unit(product.unit),
            )
            for product in recipe_request.products
        ]

    @staticmethod
    def to_recipe_response_with_products(recipe: Recipe) -> RecipeResponse:
        return RecipeResponse(
            id=recipe.id,
            name=recipe.name,
            description=recipe.description,
            servings=recipe.servings,
            products=RecipeMapper._get_recipe_product_responses(recipe),
        )

    @staticmethod
    def _get_recipe_product_responses(
        recipe: Recipe,
    ) -> list[RecipeProductResponse]:
        return [
            RecipeProductResponse(
                product=ProductMapper.to_product_response(
                    product_association.product
                ),
                quantity=product_association.quantity,
                unit=UnitMapper.to_unit(product_association.unit),
            )
            for product_association in recipe.product_associations
        ]

File: peas_frontend/src/mappers/RecipeMapper.ts
import {AddRecipe, Pagination, Recipe} from "../dto/Recipe";
import {RecipeProduct} from "../dto/RecipeProduct";
import {
  RecipeResponse,
  RecipeProductResponse,
  RecipeRequest,
  ProductDetail,
  PaginationRecipeResponse
} from "../generated/Api";
import {mapProductResponseToProduct} from "./ProductMapper";
import {mapToProductUnit, mapToUnit} from "./UnitMapper";

export function mapToPaginatedRecipeResponse(paginationRecipeResponse: PaginationRecipeResponse): Pagination<Recipe> {
  return {
    page: paginationRecipeResponse.page,
    page_size: paginationRecipeResponse.page_size,
    pages: paginationRecipeResponse.pages,
    total: paginationRecipeResponse.total,
    items: paginationRecipeResponse.items.map((recipeResponse) => mapRecipeResponseToRecipe(recipeResponse)),
  };
}


export function mapRecipeResponseToRecipe(recipeResponse: RecipeResponse): Recipe {
  return {
    id: String(recipeResponse.id),
    name: recipeResponse.name,
    description: recipeResponse.description,
    servings: recipeResponse.servings,
    products: recipeResponse.products.map((product) => mapRecipeProductResponseToRecipeProduct(product)),
  };
}

export function mapRecipeProductResponseToRecipeProduct(recipeProductResponse: RecipeProductResponse): RecipeProduct {
  return {
    product: mapProductResponseToProduct(recipeProductResponse.product),
    quantity: recipeProductResponse.quantity,
    unit: mapToProductUnit(recipeProductResponse.unit),
  };
}

export function mapRecipeProductToProductDetail(recipeProduct: RecipeProduct): ProductDetail {
  return {
    id: recipeProduct.product.id,
    quantity: recipeProduct.quantity,
    unit: mapToUnit(recipeProduct.unit),
  };
}

export function mapRecipeToRecipeRequest(recipe: Recipe): RecipeRequest {
  return {
    name: recipe.name,
    description: recipe.description,
    servings: recipe.servings,
    products: recipe.products.map((product) => mapRecipeProductToProductDetail(product)),
  };
}

export function mapAddRecipeToRecipeRequest(addRecipe: AddRecipe): RecipeRequest {
  return {
    name: addRecipe.name,
    description: addRecipe.description,
    servings: addRecipe.servings,
    products: addRecipe.products.map((product) => mapRecipeProductToProductDetail(product)),
  }
}
